packages/internals/src/utils/callOnce.ts

```diff
@@ -0,0 +1,32 @@
+type CachedValue<T> = { kind: 'empty' } | { kind: 'in-progress'; value: Promise<T> } | { kind: 'complete'; value: T }
+type AsyncFn<Args extends unknown[], R> = (...args: Args) => Promise<R>
+
+/**
+ * Takes an async function `fn` as a parameters and returns a wrapper function, which ensures
+ * that `fn` will be called only once:
+ *
+ * - If the first call is not finished yet, returns the promise to the same result
+ * - If the first call is finished, returns the result of this call
+ * @param fn
+ * @returns
+ */
+export function callOnce<Args extends unknown[], R>(fn: AsyncFn<Args, R>): AsyncFn<Args, R> {
+  let cachedValue: CachedValue<R> = { kind: 'empty' }
+  return (...args) => {
+    switch (cachedValue.kind) {
+      case 'empty':
+        cachedValue = {
+          kind: 'in-progress',
+          value: fn(...args).then((value) => {
+            cachedValue = { kind: 'complete', value }
+            return cachedValue.value
+          }),
+        }
+        return cachedValue.value
+      case 'in-progress':
+        return cachedValue.value
+      case 'complete':
+        return Promise.resolve(cachedValue.value)
+    }
+  }
+}
```

Comment:
Suggesting a smaller version that seems equivalent in terms of performance results. 
```suggestion
export function callOnce<P extends any[], R>(fn: (...args: P) => R) {
  let result: { value: R } | undefined

  return (...args: P) => {
    return result?.value ?? (result = { value: fn(...args) }).value
  }
}
```

Reply:
Thanks for suggestion, I went with slightly different variant based on it